File: submissions/cpp/90.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int> &nums) {
        vector<vector<int>> ret;
        vector<int> temp;
        int n = nums.size();
        sort(nums.begin(),nums.end());
        BackTracking(nums, ret, temp, 0, n);
        map<vector<int>, bool> exist;
        for (int i = 0; i < ret.size(); ++i) {
            if (exist.find(ret[i]) == exist.end()) {
                exist.insert(map<vector<int>, bool>::value_type(ret[i], true));
            } else {
                ret.erase(ret.begin() + i);
                --i;
            }
        }
        return ret;
    }

    void BackTracking(vector<int> &num, vector<vector<int>> &ret, vector<int> &temp, int i, int &n) {
        ret.push_back(temp);
        if (i == n)
            return;
        for (int j = i; j < n; ++j) {
            temp.push_back(num[j]);
            BackTracking(num, ret, temp, j + 1, n);
            temp.pop_back();
        }
    }
};

File: LeetCode/python3/90.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ret = []
        count = collections.Counter(nums)
        self.Backtrack([m for m in set(nums)], [], ret, 0, count)
        return ret

    def Backtrack(self, nums, curr, ret, k, count):
        ret.append([c for c in curr])
        if k == len(nums):
            return curr
        for i in range(k, len(nums)):
            temp = []
            for j in range(count[nums[i]]):
                temp.append(nums[i])
                curr += temp
                curr = self.Backtrack(nums, curr, ret, i + 1, count)
                curr = curr[:-len(temp)]
        return curr